### PR TITLE
Image limitrange

### PR DESCRIPTION
**Pull request by pweil-**
wip

### CODE REVIEW

**Comment by pweil-** [comment]
@miminar - here is my wip so far for limit ranges.  Need to spend some time with readonly containers in SCC but will get back to it asap.  Just tagging here for now so you can see where it stands.

**Comment by miminar** [comment]
Looks good :+1:

**Comment by pweil-** [comment]
@miminar @soltysh - ok, this is limiting based on image sizes in the `quotaRestrictedBlobStore` now and also catching `ImageStreamMappings` in admission in case someone is not using our internal registry images.  

Please take a look and tell me if I've missed something here or if you think this is ready to move to a real PR.

File: pkg/image/admission/admission.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	limitRanger, err := limitranger.NewLimitRanger(client, plugin.imageLimitRanger)
+	if err != nil {
+		return nil, err
+	}
+	plugin.LimitRanger = limitRanger
+
+	return plugin, nil
+}
+
+func (a *imageLimitRangerPlugin) imageLimitRanger(limitRange *kapi.LimitRange, kind string, obj runtime.Object) error {
```
Comment (pweil-) [inline pkg/image/admission/admission.go]: reminder: move to a generic function, no need to attach to `a`

File: pkg/dockerregistry/server/quotarestrictedblobstore.go [hunk 2]
```diff
@@ -126,6 +127,47 @@ func admitBlobWrite(ctx context.Context, repo *repository, size int64) error {
 	if size < 1 {
 		return nil
 	}
+
+	err := admitQuotas(ctx, repo, size)
+	if err != nil {
+		return err
+	}
+
+	err = admitLimitRanges(ctx, repo, size)
```
Comment (derekwaynecarr) [inline pkg/dockerregistry/server/quotarestrictedblobstore.go]: Should we check limits before quota?

File: pkg/image/api/types.go [hunk 3]
```diff
@@ -40,6 +40,10 @@ const (
 	// ResourceImageSize stands for a size of an image stored in an internal registry. Used as a resource
 	// quota key. If set, bigger images will be refused during a push.
 	ResourceImageSize kapi.ResourceName = "openshift.io/imagesize"
+
+	// Limit that applies to images.  Used with a max["storage"] LimitRangeItem to set
+	// the maximum size of an image.
+	LimitTypeImage kapi.LimitType = "Image"
```
Comment (derekwaynecarr) [inline pkg/image/api/types.go]: This needs to be a qname

File: pkg/image/admission/admission.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		return nil
+	}
+
+	img := ism.Image
+	if value, exists := img.Annotations[imageapi.ManagedByOpenShiftAnnotation]; !exists || value != "true" {
+		glog.V(5).Infof("image is not managed by openshift, ignoring")
+		return nil
+	}
+
+	imageSize := img.DockerImageMetadata.Size
```
Comment (miminar) [inline pkg/image/admission/admission.go]: If some older registry is in use, the `DockerImageMetadata` may not be filled yet. Use `ImageWithMetadata()` to load it from manifest.

File: pkg/image/admission/admission.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+func AdmitImage(size int64, limit kapi.LimitRangeItem) error {
+	imageQuantity := resource.NewQuantity(size, resource.BinarySI)
+
+	if limit.Type == imageapi.LimitTypeImage {
+		if limitQuantity, ok := limit.Max[kapi.ResourceStorage]; ok {
+			if limitQuantity.Cmp(*imageQuantity) < 0 {
+				// image size is larger than the permitted limit range max size, image is forbidden
+				return fmt.Errorf("Maximum %s usage per %s is %s", kapi.ResourceStorage, imageapi.LimitTypeImage, limitQuantity.String())
```
Comment (miminar) [inline pkg/image/admission/admission.go]: I'd interested to see the image size as well in the error.

File: pkg/image/admission/admission.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+			}
+		}
+	}
+	return nil
+}
+
+func (a *imageLimitRangerPlugin) Admit(attr kadmission.Attributes) error {
+	if attr.GetResource() != imageapi.Resource("imagestreammappings") || attr.GetSubresource() != "" {
```
Comment (miminar) [inline pkg/image/admission/admission.go]: `imagestream` and `imagestreamtag` need to be handled as well.

File: pkg/image/admission/admission.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	}
+	ism, ok := obj.(*imageapi.ImageStreamMapping)
+	if !ok {
+		glog.V(4).Infof("obj with type %s was sent and could not be handled", kind)
+		return nil
+	}
+
+	img := ism.Image
+	if value, exists := img.Annotations[imageapi.ManagedByOpenShiftAnnotation]; !exists || value != "true" {
+		glog.V(5).Infof("image is not managed by openshift, ignoring")
```
Comment (miminar) [inline pkg/image/admission/admission.go]: Missing image name or docker reference.

File: pkg/image/api/types.go (hunk 3 above)
Comment (miminar) [inline pkg/image/api/types.go]: Perhaps use `ResourceImageSize` directly?

**Comment by pweil-** [comment]
Closed in favor of https://github.com/openshift/origin/pull/7989

File: pkg/dockerregistry/server/quotarestrictedblobstore.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/dockerregistry/server/quotarestrictedblobstore.go]: Yes

Comment (pweil-) [inline pkg/dockerregistry/server/quotarestrictedblobstore.go]: updated in feedback commit of https://github.com/openshift/origin/pull/7989/commits/4a54ec640606eba808c536d233398ab63532db33

**Pull request closed without merging**